components/experimental/src/units/ratio.rs

```diff
@@ -195,24 +208,184 @@ impl From<u32> for IcuRatio {
-        let denominator = parts.next();
+    /// Converts a string representation of a ratio into an `IcuRatio`.
+    /// Supported string formats include:
+    /// - Fractional notation: "1/2" becomes "1/2".
+    /// - Decimal notation: "1.5" becomes "3/2".
+    /// - Scientific notation: "1.5E6" becomes "1500000", "1.5E-6" becomes "0.0000015".
+    /// - Scientific notation with commas: "1,500E6" becomes "1500000000". Commas are disregarded.
+    /// - Integer notation: "1" becomes "1".
+    /// - Empty string: "" becomes "0".
+    /// - Negative numbers: "-1/2" becomes "-1/2".
+    /// - Negative numbers with decimal notation: "-1.5" becomes "-3/2".
+    /// - Negative numbers with scientific notation: "-1.5E6" becomes "-1500000".
+    /// - Negative numbers with scientific notation with commas: "-1,500E6" becomes "-1500000000".
+    /// Parsing errors are returned for:
+    /// - Division by zero: "1/0".
+    /// - Multiple slashes: "1/2/3".
+    /// - Non-numeric characters in fractions: "1/2A".
+    /// - Multiple scientific notations: "1.5E6E6".
+    /// - Multiple decimal points: "1.5.6".
+    /// - Exponent part is not an integer: "1.5E6.5".
+    /// NOTE:
+    ///    You can add as many commas as you want in the string, they will be disregarded.
+    fn from_str(number_str: &str) -> Result<Self, IcuRatioError> {
+        /// This function interprets numeric strings and converts them into an `IcuRatio` object, supporting:
+        /// - Simple fractions: e.g., "1/2".
+        /// - Decimals: e.g., "1.5".
+        /// - Mixed fractions: e.g., "1.5/6", "1.4/5.6".
+        /// Note: Scientific notation is not supported.
+        fn parse_fraction(decimal: &str) -> Result<IcuRatio, IcuRatioError> {
+            let mut components = decimal.split('/');
+            let numerator = components.next();
+            let denominator = components.next();
+            if components.next().is_some() {
+                return Err(IcuRatioError::MultipleSlashes);
+            }
+
+            let numerator = match numerator {
+                Some(numerator) => parse_decimal(numerator)?,
+                None => IcuRatio::zero(),
+            };
+
+            let denominator = match denominator {
+                Some(denominator) => parse_decimal(denominator)?,
+                None => IcuRatio::one(),
+            };
+
+            if denominator.is_zero() {
+                return Err(IcuRatioError::DivisionByZero);
+            }
+
+            Ok(numerator / denominator)
+        }
+
+        /// Parses a decimal number from a string input.
+        /// Accepts input in various decimal formats, including:
+        /// - Whole numbers: "1"
+        /// - Decimal numbers: "1.5"
+        /// An empty string input is interpreted as "0".
+        /// Note: Fractional inputs are not supported in this context.
+        fn parse_decimal(decimal: &str) -> Result<IcuRatio, IcuRatioError> {
+            let dot_count = decimal.matches('.').count();
```

Comment:
issue: this scans the string multiple times. Instead, do. `.split('.')` and call `.next()` a few times

Reply:
done